infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+
+  @Inject
+  public AsyncStorageProvisioner(
+      @Named("che.infra.kubernetes.pvc.quantity") String pvcQuantity,
+      @Named("che.infra.kubernetes.async.storage.image") String image,
+      @Named("che.infra.kubernetes.pvc.access_mode") String accessMode,
+      @Named("che.infra.kubernetes.pvc.strategy") String strategy,
+      @Named("che.infra.kubernetes.pvc.name") String configuredPVCName,
+      @Named("che.infra.kubernetes.pvc.storage_class_name") String pvcStorageClassName,
+      SshManager sshManager,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.pvcQuantity = pvcQuantity;
+    this.storageImage = image;
+    this.accessMode = accessMode;
+    this.strategy = strategy;
+    this.configuredPVCName = configuredPVCName;
+    this.pvcStorageClassName = pvcStorageClassName;
+    this.sshManager = sshManager;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void provision(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!"true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))) {
+      return;
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !"common".equals(strategy)) {
+      String message =
+          format(
+              "Workspace configuration not valid: Asynchronous storage available only for 'common' PVC strategy, but got %s",
+              strategy);
+      LOG.warn(message);
+      osEnv.addWarning(new WarningImpl(4200, message));
+      throw new InfrastructureException(message);
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !isEphemeral(osEnv.getAttributes())) {
+      String message =
+          format("Workspace configuration not valid: Asynchronous storage available only if '%s' attribute set to false", ASYNC_PERSIST_ATTRIBUTE );
+      LOG.warn(message);
+      osEnv.addWarning(new WarningImpl(4200, message));
+      throw new InfrastructureException(message);
+    }
+
+    String namespace = identity.getInfrastructureNamespace();
+    KubernetesClient oc = clientFactory.create(identity.getWorkspaceId());
+
+    boolean isPvcExist =
+        oc.persistentVolumeClaims()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<PersistentVolumeClaim>)
+                    pvc -> pvc.getMetadata().getName().equals(configuredPVCName));
+
+    if (!isPvcExist) {
+      PersistentVolumeClaim pvc =
+          KubernetesObjectUtil.newPVC(
+              configuredPVCName, accessMode, pvcQuantity, pvcStorageClassName);
+      oc.persistentVolumeClaims().inNamespace(namespace).create(pvc);
+    }
+
+    String configMapName = namespace + ASYNC_STORAGE_CONFIG;
+    boolean isMapExist =
+        oc.configMaps()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
```

Comment:
You're enumerating all configmaps in the whole namespace to only then look for a single one. What about just trying to get it by name?

Reply:
fixed